### PR TITLE
Add feature to support multiple IDP certificate

### PR DESCRIPTION
**Pull request by nilasini**
### Proposed changes in this pull request

Resolves https://github.com/wso2/product-is/issues/3295
-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/IdentityProvider.java [hunk 1]
```diff
@@ -365,19 +379,167 @@ public void setAlias(String alias) {
...
+
+        setCertificateInfoArray(certificate);
         this.certificate = certificate;
     }
 
+    /**
+     * Get certificate info array
+     * @return CertificateInfo array
+     */
+
```
Comment (sameeragunarathne) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/IdentityProvider.java]: remove extra line

File: service-stubs/identity/org.wso2.carbon.identity.application.mgt.stub/src/main/resources/IdentityApplicationManagementService.wsdl [hunk 2]
```diff
@@ -8,34 +8,20 @@
                 </xs:sequence>
             </xs:complexType>
         </xs:schema>
-        <xs:schema xmlns:ax2139="http://common.application.identity.carbon.wso2.org/xsd" xmlns:ns="http://org.apache.axis2/xsd" xmlns:ax2143="http://model.common.application.identity.carbon.wso2.org/xsd" attributeFormDefault="qualified" elementFormDefault="qualified" targetNamespace="http://org.apache.axis2/xsd">
+        <xs:schema xmlns:ax2174="http://model.common.application.identity.carbon.wso2.org/xsd" xmlns:ns="http://org.apache.axis2/xsd" xmlns:ax2170="http://common.application.identity.carbon.wso2.org/xsd" attributeFormDefault="qualified" elementFormDefault="qualified" targetNamespace="http://org.apache.axis2/xsd">
```
Comment (sameeragunarathne) [inline service-stubs/identity/org.wso2.carbon.identity.application.mgt.stub/src/main/resources/IdentityApplicationManagementService.wsdl]: is it just a namespace reference change here?

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java [hunk 3]
```diff
@@ -78,6 +78,9 @@ private IdentityApplicationConstants(){
     public static final String ALLOW_MODIFY_USERNAME_ELEMENT = "AllowModifyUserName";
     public static final String PROMPT_CONSENT_ELEMENT = "PromptConsent";
 
+    public static final String BEGIN_CERTIFICATE = "-----BEGIN CERTIFICATE-----";
+    public static final String END_CERTIFICATE = "-----END CERTIFICATE-----";
```
Comment (madurangasiriwardena) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java]: These 2 constants were defined in IdentityProvider also. Shall we use one of these?

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationManagementUtil.java [hunk 4]
```diff
@@ -473,6 +474,49 @@ public static CertData getCertData(String encodedCert) throws CertificateExcepti
...
+                CertificateFactory factory = CertificateFactory.getInstance("X.509");
+                X509Certificate cert = (X509Certificate) factory
+                        .generateCertificate(new ByteArrayInputStream(bytes));
+                Format formatter = new SimpleDateFormat("dd/MM/yyyy");
+                certData[i] = new CertData();
+                certData[i] = fillCertData(cert, formatter);
+                i++;
+            } else {
+                String errorMsg = "Invalid encoded certificate: \'NULL\'";
+                log.debug(errorMsg);
```
Comment (madurangasiriwardena) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationManagementUtil.java]: shall we enclose the log line in "if (log.isDebugEnabled())"

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit.jsp [hunk 5]
```diff
@@ -3296,52 +3310,54 @@
                                         </tr>
                                         </thead>
                                         <tbody>
+                                        <%for(CertData certData1:certDataArr) { %>
```
Comment (madurangasiriwardena) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit.jsp]: Shall we have a better variable name than certData1

File: components/idp-mgt/org.wso2.carbon.idp.mgt/pom.xml [hunk 6]
```diff
@@ -58,6 +58,14 @@
             <groupId>org.wso2.carbon.identity.framework</groupId>
             <artifactId>org.wso2.carbon.identity.base</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.json.wso2</groupId>
+            <artifactId>json</artifactId>
+        </dependency>
+        <dependency>
```
Comment (madurangasiriwardena) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/pom.xml]: Any reason to have 2 json libraries?

Comment (nilasini) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/pom.xml]: one is json other one is gson.

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java (hunk 3 above)
Comment (nilasini) [reply components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java]: Fixed it

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/IdentityProvider.java [hunk 7]
```diff
@@ -23,16 +23,25 @@
...
+import java.nio.charset.Charset;
+import java.security.NoSuchAlgorithmException;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Set;
 
+import static org.wso2.carbon.identity.application.common.util.IdentityApplicationConstants.BEGIN_CERTIFICATE;
```
Comment (sameeragunarathne) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/IdentityProvider.java]: it's better to remove static imports on constants. You can use  IdentityApplicationConstants.BEGIN_CERTIFICATE instead on the code.

**Comment by nilasini** [comment]
Since this PR is in work in progress state. I am closing the PR. Once finish the all the suggested changes will send a PR.

**Comment by nilasini** [comment]
Open a New PR with https://github.com/wso2/carbon-identity-framework/pull/1776

**Pull request closed without merging**